unittests/test-contracts/savanna/common/savanna.hpp

```diff
@@ -0,0 +1,459 @@
+   void _check_qc(const quorum_certificate& qc, const checksum256& finality_digest, const finalizer_policy_input finalizer_policy){
+
+      auto fa_itr = finalizer_policy.finalizers.begin();
+      auto fa_end_itr = finalizer_policy.finalizers.end();
+      size_t finalizer_count = std::distance(fa_itr, fa_end_itr);
+      savanna::bitset b(finalizer_count, qc.finalizers);
+
+      bool first = true;
+
+      size_t index = 0;
+      uint64_t weight = 0;
+
+      bls_g1 agg_pub_key;
+
+      while (fa_itr != fa_end_itr){
+          if (b.test(index)){
+              bls_g1 pub_key = decode_bls_public_key_to_g1(fa_itr->public_key);
+              if (first){
+                  first=false;
+                  agg_pub_key = pub_key;
+              }
+              else agg_pub_key = _g1add(agg_pub_key, pub_key);
+              weight+=fa_itr->weight;
+          }
+          index++;
+          fa_itr++;
+      }
+
+      //verify that we have enough vote weight to meet the quorum threshold of the target policy
+      check(weight>=finalizer_policy.threshold, "insufficient signatures to reach quorum");
+      std::array<uint8_t, 32> fd_data = finality_digest.extract_as_byte_array();
+      std::string message(fd_data.begin(), fd_data.end());
+
+      std::string s_agg_pub_key = encode_g1_to_bls_public_key(agg_pub_key);
+      //verify signature validity
+      _verify(s_agg_pub_key, qc.signature, message);
+   }
+
+
+   struct authseq {
+      name account;
+      uint64_t sequence = 0;
+
+      EOSLIB_SERIALIZE( authseq, (account)(sequence) )
+
+   };
+
+   struct action_base {
+      name             account;
+      name             name;
+      std::vector<permission_level> authorization;
+
+   };
+
+   struct action :  action_base {
+      std::vector<char>    data;
+      std::vector<char>    return_value;
+
+      checksum256 digest() const {
+         checksum256 hashes[2];
+         const action_base* base = this;
+         const auto action_input_size = pack_size(data);
+         const auto return_value_size = pack_size(return_value);
+         const auto rhs_size = action_input_size + return_value_size;
+         const auto serialized_base = pack(*base);
+         const auto serialized_data = pack(data);
+         const auto serialized_output = pack(return_value);
+         hashes[0] = sha256(serialized_base.data(), serialized_base.size());
+         std::vector<uint8_t> h1_result(rhs_size);
+         std::copy (serialized_data.cbegin(), serialized_data.cend(), h1_result.begin());
+         std::copy (serialized_output.cbegin(), serialized_output.cend(), h1_result.begin() + action_input_size);
+         hashes[1] = sha256(reinterpret_cast<char*>(h1_result.data()), rhs_size);
+         return hash_pair(std::make_pair(hashes[0], hashes[1]));
+      }
+
+      EOSLIB_SERIALIZE( action, (account)(name)(authorization)(data)(return_value))
+
+   };
+
+   struct action_data {
+

```

Comment:
Can empty lines after beginning of a struct and before end of a struct be removed? They seem excessive.